content/posts/creating-grafana-dashboards-for-nodejs-on-kubernetes/index.md

```diff
@@ -0,0 +1,261 @@
+
+```
+# HELP http_request_duration_seconds duration histogram of http responses labeled with: status_code, method
+# TYPE http_request_duration_seconds histogram
+http_request_duration_seconds_bucket{le="0.003",status_code="200",method="GET"} 1204
+http_request_duration_seconds_bucket{le="0.03",status_code="200",method="GET"} 1214
+http_request_duration_seconds_bucket{le="0.1",status_code="200",method="GET"} 1215
+http_request_duration_seconds_bucket{le="0.3",status_code="200",method="GET"} 1215
+http_request_duration_seconds_bucket{le="1.5",status_code="200",method="GET"} 1230
+http_request_duration_seconds_bucket{le="10",status_code="200",method="GET"} 1257
+http_request_duration_seconds_bucket{le="+Inf",status_code="200",method="GET"} 1260
+http_request_duration_seconds_sum{status_code="200",method="GET"} 159.43762229599974
+http_request_duration_seconds_count{status_code="200",method="GET"} 1260
+.
+.
+.
+
+# HELP up 1 = up, 0 = not up
+# TYPE up gauge
+up 1
+```
+
+You can see, in this case the endpoint exposes multiple metrics called `http_request_duration_seconds_bucket` with different values for the `le` label. These are duplicated for other status codes, but those have been omitted in this example. When Prometheus scrapes the endpoint, it will add additional labels to provide metadata for later queries. Prometheus features built-in service discovery options for adding Kubernetes metadata such as the namespace of the pod, the pod name, and Kubernetes labels assigned to the pod.
+
+The `http_request_duration_seconds_count` metric from above may look something like this after scraping:
+
+```
+http_request_duration_seconds_count{app="awesome-node", instance="10.244.2.130:8080", job="kubernetes-pods", kubernetes_namespace="awesome-node", kubernetes_pod_name="awesome-node-56bf9dfd49-74v5x", method="GET", pod_template_hash="56bf9dfd49", status_code="200"}
+```
+
+### Libraries for NodeJS
+
+There are a number of libraries available for NodeJS that export some basic metrics for many NodeJS webservers.
+
+In this example we're going to use the [Express Prometheus Bundle](https://github.com/jochen-schweizer/express-prom-bundle), which also supports Koa, and exposes metrics very suitable for the Prometheus data model and query language.
+
+The Express Prometheus Bundle automatically exports a number of useful metrics related to request duration and is based on [prom-client](https://github.com/siimon/prom-client) which you can use to export additional custom metrics for your application.
+
+## Example Application
+
+We're working with bespinian's Express based [Awesome Node](https://github.com/bespinian/awesome-node) example application. It's deployed into a namespace called `awesome-node`, exports metrics using the Express Prometheus Bundle and is running 3 replicas.
```

Comment:
I agree. How about something like this:

In this example, we're working with bespinian's Express based [Awesome Node](https://github.com/bespinian/awesome-node) example application. It's deployed into a namespace called `awesome-node`, exports metrics using the Express Prometheus Bundle and is running 3 replicas.

If your application already exposes metrics and is set up in your Kubernetes cluster, follow along and implement the dashboard as we go. Otherwise it may be a good exercise to deploy the Awesome Node application and set up the sample dashboard using the data it generates.